shared/js/intl_helper.js

```diff
@@ -0,0 +1,170 @@
+(function(global) {
+  'use strict';
+
+  /*************************
+
+  IntlHelper.define('shortDate', 'datetime', {
+    year: 'numeric',
+    month: 'long',
+  });
+
+  function setValue() {
+    var formatter = IntlHelper.get('shortDate');
+    console.log(formatter.format(new Date()));
+  }
+
+  setValue();
+
+  IntlHelper.observe('shortDate', setValue);
+  
+   ****************************/
+  var objectCache;
+
+  const knownObjects = {
+    'datetime': {
+      set: function(options) {
+        const customOptions = Object.assign({}, options);
+        if (options.hour) {
+          customOptions.hour12 = navigator.mozHour12;
+        }
+        return Intl.DateTimeFormat(
+          navigator.languages,
+          customOptions
+        );
+      },
+      isAffected: function(obj, reason) {
+        if (reason !== 'timeformatchange') {
+          return true;
+        }
+
+        return 'hour' in obj.options;
+      }
+    },
+    'number': {
+      set: function(options) {
+        return Intl.NumberFormat(
+          navigator.languages,
+          options
+        );
+      },
+      isAffected: function(obj, reason) {
+        return true;
+      }
+    },
+    'collator': {
+      set: function(options) {
+        return Intl.Collator(
+          navigator.languages,
+          options
+        );
+      },
+      isAffected: function(obj, reason) {
+        return true;
+      }
+    },
+  };
+
+  const affectedTypes = {
+    'languagechange': ['datetime', 'number', 'collator'],
+    'timeformatchange': ['datetime'],
+  };
+
+  global.IntlHelper = {
+    define: function(name, type, options) {
+      if (!knownObjects.hasOwnProperty(type)) {
+        throw new Error('Unknown type: ' + type);
+      }
+
+      if (!objectCache) {
+        objectCache = new Map();
+      }
+
+      if (objectCache.has(name)) {
+        throw new Error(
+          'Intl Object with name "' + name + '" already exists');
+      }
+
+      objectCache.set(name, {
+        options,
+        type,
+        object: undefined,
+        listeners: new Set()
+      });
+    },
+
+    get: function(name) {
+      if (!objectCache.has(name)) {
+        throw new Error('Intl Object with name "' + name + '" is not defined');
+      }
+      const obj = objectCache.get(name);
+      
+      if (!knownObjects.hasOwnProperty(obj.type)) {
+        throw new Error('Unknown type: ' + obj.type);
+      }
+
+      if (obj.object === undefined) {
+        obj.object = knownObjects[obj.type].set(obj.options);
```

Comment:
I thought `knownObjects[obj.type]` was a `Set` instance because of the `.set` method. Maybe rename it to something less ambiguous (`create`?).